examples/cluster-with-resources/README.md

```diff
@@ -0,0 +1,70 @@
+# Provision a cluster on AKS and install Kubernetes manifests
+
+This example demonstrates how to use the provider together with a cluster provisioned on Azure's AKS.
+
+**It's important to be aware that this example will not work as expected when applied with Terraform as a single operation.**
+
+This is because the provider requires access to the Kubernetes API during the planning phase. Initially, before any resources have been created, there is no available API endpoint and the provider will throw an error.
+
+This example demonstrates how to group resources related to the cluster in their own module and the Kubernetes manifests in another separate module, to facilitate building in distinct stages.
+
+It is still possible to build this configuration and obtain a single state file. For this, two Terraform operations are necessary.
+
+Before anything, make sure your environment is configured with valid Azure credentials.
+In paricular, have these environment variables set to relevant values in your Azure account:
+```
+ARM_SUBSCRIPTION_ID￼￼
+ARM_TENANT_ID￼￼
+ARM_CLIENT_ID (a.k.a. App ID)
+ARM_CLIENT_SECRET (a.k.a. password)
+```
+
+Initialize the workspace:
+
+```shell
+ » terraform init
+```
+
+The first Terraform apply operation will build just the AKS cluster and other resources it may require. For this, we use the `-target` argument to Terraform to limit the scope of the apply.
+
+```shell
+ » terraform apply -target module.cluster
+```
+
+Once this operation succeeds, the AKS cluster is available and the state file contains its attribute values.
+
+Running `terraform state list` should show the following resources present in the state:
+
+```shell
+ » terraform state list
+random_id.uniq
+module.cluster.azurerm_kubernetes_cluster.test
+module.cluster.azurerm_resource_group.test
+module.cluster.local_file.kubeconfig
+```
+
```

Comment:
My terraform state doesn't seem to include `random_id.uniq`, even though the first apply was successful.

```
$ terraform0.13.0-beta3 state list
module.cluster.azurerm_kubernetes_cluster.test
module.cluster.azurerm_resource_group.test
module.cluster.local_file.kubeconfig
```

During the second apply, it fails with this error:

```
$  terraform0.13.0-beta3 apply
module.cluster.azurerm_resource_group.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourceGroups/k8s-alpha-test]
module.cluster.azurerm_kubernetes_cluster.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourcegroups/k8s-alpha-test/providers/Microsoft.ContainerService/managedClusters/k8s-alpha-aks]
module.cluster.local_file.kubeconfig: Refreshing state... [id=47a9c0024202ef5808f847adb099170823cfc8d8]

Error: rpc error: code = Unknown desc = value is not known
```

I think it's because the random_id is unknown. A destroy is also unsuccessful, unless I specifically target the cluster module.

```
[dakini@dax cluster-with-resources ((b764425...) %)]$ terraform0.13.0-beta3 destroy
module.cluster.azurerm_resource_group.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourceGroups/k8s-alpha-test]
module.cluster.azurerm_kubernetes_cluster.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourcegroups/k8s-alpha-test/providers/Microsoft.ContainerService/managedClusters/k8s-alpha-aks]
module.cluster.local_file.kubeconfig: Refreshing state... [id=47a9c0024202ef5808f847adb099170823cfc8d8]

Error: rpc error: code = Unknown desc = value is not known


$  terraform0.13.0-beta3 destroy -target module.cluster
module.cluster.azurerm_resource_group.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourceGroups/k8s-alpha-test]
module.cluster.azurerm_kubernetes_cluster.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourcegroups/k8s-alpha-test/providers/Microsoft.ContainerService/managedClusters/k8s-alpha-aks]
module.cluster.local_file.kubeconfig: Refreshing state... [id=47a9c0024202ef5808f847adb099170823cfc8d8]

...

  # module.cluster.azurerm_kubernetes_cluster.test will be destroyed
  - resource "azurerm_kubernetes_cluster" "test" {
      - api_server_authorized_ip_ranges = [] -> null
      - dns_prefix                      = "k8s-alpha-aks" -> null
      - enable_pod_security_policy      = false -> null
      - fqdn                            = "k8s-alpha-aks-732b747b.hcp.westeurope.azmk8s.io" -> null
      - id                              = "/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourcegroups/k8s-alpha-test/providers/Microsoft.ContainerService/managedClusters/k8s-alpha-aks" -> null
      - kube_admin_config               = [] -> null
      - kube_config                     = [
          - {
...
```

I was going to post the trace log, but it contains secrets. I tried working around the issue by targeting the manifests module next, but that didn't work either.

```
$  terraform0.13.0-beta3 apply -target module.manifests
module.cluster.azurerm_resource_group.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourceGroups/k8s-alpha-test]
module.cluster.azurerm_kubernetes_cluster.test: Refreshing state... [id=/subscriptions/88720cb0-d9d7-4d5f-917b-7ba469118dbc/resourcegroups/k8s-alpha-test/providers/Microsoft.ContainerService/managedClusters/k8s-alpha-aks]
module.cluster.local_file.kubeconfig: Refreshing state... [id=47a9c0024202ef5808f847adb099170823cfc8d8]

Error: rpc error: code = Unknown desc = value is not known
```
I think I'm stuck on this... at least for now.

Reply:
This looks like the useful part of the trace log.

```
2020/08/07 14:26:34 [TRACE] EvalReadState: reading state for module.manifests.kubernetes_manifest.test-cfm
2020/08/07 14:26:34 [TRACE] EvalReadState: no state present for module.manifests.kubernetes_manifest.test-cfm
2020/08/07 14:26:34 [TRACE] eval: *terraform.EvalDiff
2020/08/07 14:26:34 [TRACE] Re-validating config for "module.manifests.kubernetes_manifest.test-cfm"
2020/08/07 14:26:34 [TRACE] GRPCProvider: ValidateResourceTypeConfig
2020/08/07 14:26:34 [TRACE] GRPCProvider: PlanResourceChange
2020/08/07 14:26:34 [ERROR] eval: *terraform.EvalDiff, err: rpc error: code = Unknown desc = value is not known
2020/08/07 14:26:34 [ERROR] eval: *terraform.EvalSequence, err: rpc error: code = Unknown desc = value is not known
2020/08/07 14:26:34 [TRACE] [walkRefresh] Exiting eval tree: module.manifests.kubernetes_manifest.test-cfm
2020/08/07 14:26:34 [TRACE] vertex "module.manifests.kubernetes_manifest.test-cfm": visit complete
2020/08/07 14:26:34 [TRACE] vertex "module.manifests.kubernetes_manifest.test-cfm": dynamic subgraph encountered errors
2020/08/07 14:26:34 [TRACE] vertex "module.manifests.kubernetes_manifest.test-cfm": visit complete
2020/08/07 14:26:34 [TRACE] vertex "module.manifests.kubernetes_manifest.test-cfm (expand)": dynamic subgraph encountered errors
2020/08/07 14:26:34 [TRACE] vertex "module.manifests.kubernetes_manifest.test-cfm (expand)": visit complete
2020/08/07 14:26:34 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/kubernetes-alpha\"] (close)" errored, so skipping
2020/08/07 14:26:34 [TRACE] dag/walk: upstream of "module.manifests (close)" errored, so skipping
2020/08/07 14:26:34 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/08/07 14:26:34 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
ESC[31m
2020/08/07 14:26:34 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
ESC[1mESC[31mError: ESC[0mESC[0mESC[1mrpc error: code = Unknown desc = value is not knownESC[0m
```

Reply:
I got it working! Here are the changes I made. You can use them if you want, or I can push to the branch. Basically the random ID stuff wasn't successfully being passed into the manifests module, I think. So I moved the random_id into the cluster module to use it in both modules.  
https://gist.github.com/dak1n1/0e52587a96423d7f47ab0f22cf2b8cf3

Reply:
oops, I spoke too soon. Both applies work now, but the `destroy` does not:

```
Error: Cycle: module.cluster.output.host (expand), module.cluster.output.client_certificate (expand), module.cluster.output.cluster_ca_certificate (expand), provider["registry.terraform.io/hashicorp/kubernetes-alpha"], module.manifests.kubernetes_manifest.test-cfm (destroy), module.cluster.azurerm_kubernetes_cluster.test (destroy), module.cluster.output.client_key (expand)
```

It only works after I manually remove the manifest module from my state file.